### DISCUSSION TITLE
Unable to attach a  debugger in VS Code

### DISCUSSION THREAD

**Discussion by jeantoman** [vscode-abl/vscode-abl#248] [Q&A] [answered]
Hello, 

First of all I'd like to thank you for a great project.

We're facing an issue with attaching a debugger to the local project in Progress 12.2. 

If I am running the code using **ABL: Run prowin** or **ABL: run _progress -b** everyting works. Connection to the Db is working as well.

Using Start debugging (F5) or even Run without debugging causes Internal error in VS. 

I've tried several launch configurations, create the openedge-project.json with provided code here on the page, launch.jsons etc.

Thanks alot!

![image](https://github.com/user-attachments/assets/8d317352-4245-4261-b7b4-72b3dd4ce28b)

![image](https://github.com/user-attachments/assets/510d82ee-688d-41c6-bff4-879292b1d568)

**Comment by gquerret** [comment]
See [this comment](https://github.com/vscode-abl/vscode-abl/issues/235#issuecomment-2409901728).

**Reply by jeantoman** [reply]
Thank you for a quick response, I' ve tried to attach 9999, unfortunately port 9999 is occupied in our vm, so i tried to attach 9998 with attached launch settings but without success. Still the same error.
![image](https://github.com/user-attachments/assets/a701f9e5-cd0b-42fe-9db1-f5bc183abb72)

**Reply by gquerret** [reply]
You're trying to connect hostname `0.0.0.0`, that probably should be `127.0.0.1`. The `mode` and `pause` attributes are missing. Here is my default config:
```json
{
      "name": "Attach to AVM",
      "type": "abl",
      "request": "attach",
      "hostname": "127.0.0.1",
      "mode": "legacy",
      "port": 3099,
      "localRoot": "${workspaceFolder}",
      "pause": true
    }
```

**Reply by jeantoman** [reply]
still the same error, even if i release 9999 port, still getting this internal error without any information what could be wrong

**Reply by jeantoman** [reply]
Here is my openedge-project.json 
![image](https://github.com/user-attachments/assets/e0edbc03-b6f7-40cd-ac59-b1fa22fad83d)

**Reply by gquerret** [reply]
Perhaps a stupid question, but do you have an AVM listening on this port number before starting the debugger?
![image](https://github.com/user-attachments/assets/ce41fb2a-00f3-4260-944b-872e9a2aeeb5)

Then the debug console should also provide basic information:
![image](https://github.com/user-attachments/assets/20128e5c-8be5-45c6-877e-7bda488f56c4)

And ultimately, there's a log file (name starting with `vscode-debug-adapter`) which is generated in the %TEMP% directory of the user. Note that the current stable version of the plugin has an issue (file is not generated), so you should switch to the prerelease stream.
![image](https://github.com/user-attachments/assets/c25b4050-7003-48c8-aa31-5864b6e85539)

**Reply by jeantoman** [reply]
all done, same error. I tried to run app on 9999 as well as 3099

**Reply by jeantoman** [reply]
1) warning appears on every run **prowin <program location> -debugready XXXX** 
2) i see anything there under the debug label
3) I tried both of them - pre and rel versions. 

log from temp:

[2024-11-23T18:28:54,455] [INFO] Client connected
[2024-11-23T18:28:54,493] [INFO] Initialize Adapter
[2024-11-23T18:28:54,522] [INFO] Attach
[2024-11-23T18:28:54,523] [INFO]  --> Argument 'name': 'Attach to AVM' (java.lang.String)
[2024-11-23T18:28:54,525] [INFO]  --> Argument 'type': 'abl' (java.lang.String)
[2024-11-23T18:28:54,525] [INFO]  --> Argument 'request': 'attach' (java.lang.String)
[2024-11-23T18:28:54,525] [INFO]  --> Argument 'hostname': '127.0.0.1' (java.lang.Strin

......some propath stuf etc.....

[2024-11-23T18:28:55,545] [INFO] [projectName] Read typeInfoRoot cache in 1 ms
[2024-11-23T18:28:55,545] [INFO] [projectName] Read propathTypeInfoRoot cache in 0 ms
[2024-11-23T18:28:55,545] [INFO] [projectName Read catalog cache in 0 ms
[2024-11-23T18:28:55,547] [INFO] [projectName] TypeInfo cache cleanup in 1 ms

**Reply by gquerret** [reply]
Can you trying connecting the legacy debugger to your AVM ? I'd like to make sure this is a debug adapter problem or not.
And what do you see in the debug console ?

<details>
<summary>Sample output with successful connection</summary>
[INFO] Client connected
[INFO] Initialize Adapter
[INFO] Attach
[INFO]  --> Argument 'name': 'Attach to AVM' (java.lang.String)
[INFO]  --> Argument 'type': 'abl' (java.lang.String)
[INFO]  --> Argument 'request': 'attach' (java.lang.String)
[INFO]  --> Argument 'hostname': '127.0.0.1' (java.lang.String)
[INFO]  --> Argument 'mode': 'legacy' (java.lang.String)
[INFO]  --> Argument 'port': '3099.0' (java.lang.Double)
[INFO]  --> Argument 'localRoot': 'C:\Users\GillesQUERRET\projets\sonaroe-test-project' (java.lang.String)
[INFO]  --> Argument 'pause': 'true' (java.lang.Boolean)
[INFO]  --> Argument '__configurationTarget': '6.0' (java.lang.Double)
[INFO]  --> Argument 'ablsrc': 'c:\Users\GillesQUERRET\.vscode\extensions\riversidesoftware.openedge-abl-lsp-1.15.1-win32-x64\resources\abl-src' (java.lang.String)
[INFO]  --> Argument 'dlc': 'C:\Progress\OpenEdge-12.8' (java.lang.String)
[INFO]  --> Argument 'oeversion': '12.8' (java.lang.String)
[INFO]  --> Argument '__sessionId': '2059139b-5b94-4ae3-a2a8-3e247db62493' (java.lang.String)
[INFO] [Test Project] Current profile: default
[INFO] [Test Project] Progress version : 12.8 -> C:\Progress\OpenEdge-12.8
[INFO] [Test Project] Source directory C:\Users\GillesQUERRET\projets\sonaroe-test-project\src\procedures
[INFO] [Test Project]   --> Build into C:\Users\GillesQUERRET\projets\sonaroe-test-project\build
[INFO] [Test Project]   --> PCT dir    C:\Users\GillesQUERRET\projets\sonaroe-test-project\.builder\.pct0
[INFO] [Test Project] Source directory C:\Users\GillesQUERRET\projets\sonaroe-test-project\src\classes
[INFO] [Test Project]   --> Build into C:\Users\GillesQUERRET\projets\sonaroe-test-project\build
[INFO] [Test Project]   --> PCT dir    C:\Users\GillesQUERRET\projets\sonaroe-test-project\.builder\.pct1
[INFO] [Test Project] PROPATH recap:
[INFO] [Test Project]   src\procedures -> C:\Users\GillesQUERRET\projets\sonaroe-test-project\src\procedures
[INFO] [Test Project]   src\classes -> C:\Users\GillesQUERRET\projets\sonaroe-test-project\src\classes
[INFO] [Test Project]   C:\Progress\OpenEdge-12.8\tty\netlib\OpenEdge.Net.pl -> C:\Progress\OpenEdge-12.8\tty\netlib\OpenEdge.Net.pl
[INFO] [Test Project]   C:\Progress\OpenEdge-12.8\tty\OpenEdge.Core.pl -> C:\Progress\OpenEdge-12.8\tty\OpenEdge.Core.pl
[INFO] [Test Project] Static init done
[INFO] Starting legacy debugger
[INFO] startSockets
[INFO] Starting reader thread
[INFO] Add message to queue: SETPROP IDE 1
[INFO] Add message to queue: interrupt
[INFO] [Test Project] Read typeInfoRoot cache in 5 ms
[INFO] [Test Project] Read propathTypeInfoRoot cache in 16 ms
[INFO] [Test Project] Read catalog cache in 0 ms
[INFO] [Test Project] TypeInfo cache cleanup in 21 ms 
</details>

**Reply by gquerret** [reply]
And can you confirm that the project configuration is correct, i.e. the output in the ABL Language Server tab doesn't show any error ?

**Comment by jeantoman** [comment]
I have no output on debug console. From ABL Language server im getting this exception 
![image](https://github.com/user-attachments/assets/3d0835fb-9029-4624-8b0a-3fe7f0176e7a)

**Reply by jeantoman** [reply]
anhother error from ABL Language server 
![image](https://github.com/user-attachments/assets/99de8275-706f-49f8-a29f-60fffb9f57b3)

**Reply by gquerret** [reply, marked as answer]
I see that the debug adapter silently swallows exception during the project initialization, that's why you don't see errors when starting the debug adapter. That will be fixed.
However, project initialization is almost the same in the language server and debug adapter, so if you fix LS startup, then you'll probably be able to fix the DA startup. Your first error can be easily fixed, just point to an existing file, or remove the documentation attribute. The second one is probably coming from your propath pointing to `${DLC}/bin` or `${DLC}`, which leads to the AccessDeniedException when searching for rcode / PL. Remove this entry.

**Comment by jeantoman** [comment]
so i got rid of all errors mentioned in a ABL LS and it has started to work :) 

cool, thank you very much! 


![image](https://github.com/user-attachments/assets/476d298a-83a8-46de-8cd6-5eeecf81569f)

**Reply by jeantoman** [reply]
If you'll visit czech republic, you'll have a shot on me, thank you for your time :)

**Reply by gquerret** [reply]
👍
